### PR TITLE
[MODULAR] Rave Cube

### PR DESCRIPTION
**Pull request by lessthnthree**
## About The Pull Request
We can't use the jukebox on Skyrat because it brings the server to its knees with our population. However, that doesn't mean the crew should be depraved of a party!

Ordered through the cargo department, the rave cube has coloured spotlights and sparkles, but no jukebox component. Includes a dance mode toggle to autodance within 4 tiles of it.

Are you a terrible dancer and don't want to join in? You can disable dancing in game preferences on a per client basis.

## How This Contributes To The Skyrat Roleplay Experience
Throw wicked parties without the server-killing jukebox components.

## Proof of Testing
<details>

https://user-images.githubusercontent.com/83487515/212588923-2a882faa-d85e-42d7-b058-0eab78514722.mp4

https://user-images.githubusercontent.com/83487515/212589397-3e4ee902-b5b5-4ef4-9ccf-91cafcb4dfd7.mp4

https://user-images.githubusercontent.com/83487515/212589414-d60246f1-cbf9-40a0-9727-d14974a01158.mp4

https://user-images.githubusercontent.com/83487515/212589448-3e8df467-39f6-42c7-badc-a6498d3437e2.mp4

</details>

## Changelog

:cl: LT3
add: RAVE CUBE is now available to be ordered through cargo. Party hard!
/:cl:

### CODE REVIEW

**Comment by Zergspower** [comment]
giphy

File: modular_skyrat/modules/rave_cube/code/rave_cube.dm [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	. = ..()
+	if(!anchored)
+		. += span_notice("<b>Alt+Click</b> to lock the rave cube in place.")
+	else
+		. += span_notice("<b>Alt+Click</b> to unlock the rave cube for travel.")
+
+/obj/machinery/rave_cube/AltClick(mob/living/carbon/human/user)
+	. = ..()
+	set_anchored(!anchored)
+	to_chat(user, span_notice("You [anchored ? null : "un"]lock the rave cube."))
```
Comment (tf-4) [inline modular_skyrat/modules/rave_cube/code/rave_cube.dm]:
```suggestion
	balloon_alert(user, "[anchored ? null : "un"]anchored"))
```

File: modular_skyrat/modules/rave_cube/code/rave_cube.dm [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+		turn_off(user)
+	else
+		turn_on(user)
+
+/obj/machinery/rave_cube/examine(mob/user)
+	. = ..()
+	if(!anchored)
+		. += span_notice("<b>Alt+Click</b> to lock the rave cube in place.")
+	else
+		. += span_notice("<b>Alt+Click</b> to unlock the rave cube for travel.")
```
Comment (tf-4) [inline modular_skyrat/modules/rave_cube/code/rave_cube.dm]:
```suggestion
	. += span_notice("<b>Alt+Click</b> to [anchored ? null : "un"]anchor [src].") 
```

File: modular_skyrat/modules/rave_cube/code/rave_cube.dm [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+		luminosity = 1
+		SSvis_overlays.add_vis_overlay(src, icon, "active", layer, plane, dir, alpha)
+		SSvis_overlays.add_vis_overlay(src, icon, "active", 0, EMISSIVE_PLANE, dir, alpha)
+
+/obj/item/flashlight/spotlight/rave
+	name = "rave light"
+	desc = "Fits in with your glowing cat ears headband. Or your actual ears if you're a feline, I suppose."
+
+/obj/machinery/rave_cube/proc/lights_setup()
+	var/turf/cen = get_turf(src)
```
Comment (tf-4) [inline modular_skyrat/modules/rave_cube/code/rave_cube.dm]:
```suggestion
	var/turf/centre = get_turf(src)
```

File: modular_skyrat/modules/rave_cube/code/rave_cube.dm [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+			spotlights += new /obj/item/flashlight/spotlight/rave(turf, 1.4 + get_dist(src, turf), 30 - (get_dist(src, turf) * 8), LIGHT_COLOR_BLUEGREEN)
+			continue
+		if((turf.x+1 == cen.x && turf.y-1 == cen.y) || (turf.x+2==cen.x && turf.y-2 == cen.y))
+			spotlights += new /obj/item/flashlight/spotlight/rave(turf, 1.4 + get_dist(src, turf), 30 - (get_dist(src, turf) * 8), LIGHT_COLOR_BLUE)
+			continue
+		continue
+	FOR_DVIEW_END
+
+/obj/machinery/rave_cube/proc/lights_spin()
+	visible_message(span_info("Rave cube is initialising... Please wait."))
```
Comment (tf-4) [inline modular_skyrat/modules/rave_cube/code/rave_cube.dm]: Could have the cube speak instead for something a little more flavourful

**Review by tf-4 — CHANGES_REQUESTED** [review]
Wrote the suggestions on my phone but you get the idea

**Comment by Nerev4r** [comment]
It would be preferable to lock the goofy-ass dances like spinning around in the air upside down and such, behind a skillchip or perhaps the Freerunning quirk.

**Comment by ghost** [comment]
Is this sprite just the blackbox but... taller and more colorful..?

**Comment by lessthnthree** [comment]
Yes.

**Comment by lessthnthree** [comment]
Turns out the long ass helicopter flipping and such was my mistake. I severely reduced the length for each of them.

**Pull request closed without merging**